cartofacade/cgo_api.go

```diff
@@ -0,0 +1,401 @@
+type GetPositionHolder struct {
+	x float64
+	y float64
+	z float64
+
+	ox    float64
+	oy    float64
+	oz    float64
+	theta float64
+
+	real float64
+	imag float64
+	jmag float64
+	kmag float64
+
+	compReference string
+}
+
+// GetInternalStateResponse holds the c type viam_carto_get_internal_state_response
+type GetInternalStateResponse struct {
+	value C.viam_carto_get_internal_state_response
+}
+
+// LidarConfig represents the lidar configuration
+type LidarConfig int64
+
+const (
+	twoD LidarConfig = iota
+	threeD
+)
+
+// CartoConfig contains config values from app
+type CartoConfig struct {
+	sensors            []string
+	mapRateSecond      int
+	dataDir            string
+	componentReference string
+	lidarConfig        LidarConfig
+}
+
+// CartoAlgoConfig contains config values from app
+type CartoAlgoConfig struct {
+	optimizeOnStart      bool
+	optimizeEveryNNodes  int
+	numRangeData         int
+	missingDataRayLength float64
+	maxRange             float64
+	minRange             float64
+	maxSubmapsToKeep     int
+	freshSubmapsCount    int
+	minCoveredArea       float64
+	minAddedSubmapsCount int
+	occupiedSpaceWeight  float64
+	translationWeight    float64
+	rotationWeight       float64
+}
+
+// NewCartoLib calls viam_carto_lib_init and returns a pointer to a viam carto lib object.
+func NewCartoLib(miniloglevel, verbose int) (CartoLib, error) {
+	pVcl := C.alloc_viam_carto_lib()
+	defer C.free_alloc_viam_carto_lib(pVcl)
+
+	ppVcl := (**C.viam_carto_lib)(unsafe.Pointer(&pVcl))
+	/*
+		Question: what is going on here? are we not getting correct amount of indirection?
+					fmt.Printf("%#v\n", pVcl) = &cartofacade._Ctype_struct_viam_carto_lib{minloglevel:0, verbose:0}
+					fmt.Println(&pVcl) = 0x14000128078
+					fmt.Println(goPtr) = 0x600001ee8000
+					// ppVcl := (**C.viam_carto_lib)(goPtr)
+					fmt.Println(ppVcl) = 0x600001ee8000
+	*/
+
+	status := C.viam_carto_lib_init(ppVcl, C.int(miniloglevel), C.int(verbose))
+	if err := getErrorFromStatusCode(status); err != nil {
+		return CartoLib{}, err
+	}
+
+	vcl := CartoLib{value: (*ppVcl)}
+
+	return vcl, nil
+}
+
+// TerminateCartoLib calls viam_carto_lib_terminate to clean up memory for viam carto lib.
+func (vcl *CartoLib) TerminateCartoLib() error {
+	status := C.viam_carto_lib_terminate((**C.viam_carto_lib)(unsafe.Pointer(vcl)))
+	if err := getErrorFromStatusCode(status); err != nil {
+		return err
+	}
+	return nil
+}
+
+// NewCarto calls viam_carto_init and returns a pointer to a viam carto object.
+func NewCarto(cfg CartoConfig, acfg CartoAlgoConfig, vcl CartoLib) (Carto, error) {
+	pVc := C.alloc_viam_carto()
+	defer C.free_alloc_viam_carto(pVc)
+
+	ppVc := (**C.viam_carto)(unsafe.Pointer(&pVc))
+
+	vcc := getConfig(cfg)
+	defer C.free_bstring_array(vcc.sensors)
+
+	vcac := getAlgoConfig(acfg)
+
+	status := C.viam_carto_init(ppVc, vcl.value, vcc, vcac)
+	if err := getErrorFromStatusCode(status); err != nil {
+		return Carto{}, err
+	}
+
+	return Carto{value: (*ppVc)}, nil
```

Comment:
Why are we storing whatppVcl points to, rather than ppVc itself?

Reply:
This is the type that is needed for all operations aside from initialization so I thought it would be best to store this instead of getting what ppVcl points to each time we want to call any method. What do you think?

Reply:
I think terminate also takes a ppVcl

Reply:
right, so for only terminate I would get a reference to this obj instead of getting what this obj points to for each other method. Happy to change it either way to what you think is best

Reply:
This works. I realized it would be more complex to also need to carry around the ppVc itself b/c we would also need to remember to free both pointers when we terminate. Your implementation here is best.